src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameLexer.cs

```diff
@@ -0,0 +1,368 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.StackFrame
+{
+    using StackFrameToken = EmbeddedSyntaxToken<StackFrameKind>;
+    using StackFrameTrivia = EmbeddedSyntaxTrivia<StackFrameKind>;
+
+    internal struct StackFrameLexer
+    {
+        public readonly VirtualCharSequence Text;
+        public int Position;
+
+        public StackFrameLexer(string text)
+            : this(VirtualCharSequence.Create(0, text))
+        {
+        }
+
+        public StackFrameLexer(VirtualCharSequence text) : this()
+            => Text = text;
+
+        public VirtualChar CurrentChar => Position < Text.Length ? Text[Position] : default;
+
+        public VirtualCharSequence GetSubPatternToCurrentPos(int start)
+            => GetSubPattern(start, Position);
+
+        public VirtualCharSequence GetSubPattern(int start, int end)
+            => Text.GetSubSequence(TextSpan.FromBounds(start, end));
+
+        public StackFrameToken ScanNextToken(bool allowTrivia)
+        {
+            var trivia = ScanLeadingTrivia(allowTrivia);
+            if (Position == Text.Length)
+            {
+                return CreateToken(StackFrameKind.EndOfLine, trivia, VirtualCharSequence.Empty);
```

Comment:
oh, you have an EOL token.  that affects https://github.com/dotnet/roslyn/pull/56957/files#r722727207

Also, why does the CompilationUnit not point at teh EOL token?